### PR TITLE
Site Editor: Fix template resolution for templates assigned as home page

### LINKED ISSUE #56267: Site Editor: Static homepage not connected to the Front Page template
Discovered in https://github.com/WordPress/gutenberg/issues/56244

The Front Page template will always resolve to display the homepage, regardless of whether reading settings dictate that it should display latest posts or a static page (see the [template hierarchy](https://developer.wordpress.org/files/2014/10/Screenshot-2019-01-23-00.20.04.png) for more info). 

However, if you set a static homepage, and Front Page exists, the Site Editor doesn't make this connection, and even tells you that an incorrect template is used when you view and edit that page. The post editor works as expected. 

Here's a video that hopefully demonstrates the issue clearly:

https://github.com/WordPress/gutenberg/assets/846565/709682b4-2c74-4e44-80f5-eeff5f11d114

To reproduce:

1. In reading settings select a static homepage
2. In the Site editor create a Front Page template (if one doesn't already exist)
3. Navigate to Pages > [your static homepage] and notice that Front Page isn't listed as the template in the details panel
4. Edit [your static homepage] and notice that Front Page isn't listed as the template in the Inspector

### PR DESCRIPTION
**Pull request by youknowriad**
closes #56267 
I think this is a regression that was introduced in #51477

## What

If a page was set as "static home page", and you had a "front page" template on your theme, this template was not being picked as the template to use for your home page, instead we were always using the default "page" template regardless of whether that template exited.

This PR fixes that by avoiding the use of `getDefaultTemplate` selector and instead uses `__experimentalGetTemplateForLink` to resolve templates for specific pages if they don't have a fixed or edited "template" property defined.

## Why?

The template resolution mechanism is a bit complex and the previous implement was trying to reimplement it entirely locally, there's another solution where we could look at the presence or not of "front page" template before falling back to `getDefaultTemplate` but I thought that was too complex since we already had a way to ask the server for the fallback template.

## Testing Instructions

1- Create a random page and use it as "static home page" in your "reading" settings
2- Ensure that you have a front page template on your theme
3- Open the site editor: The front page template should be loaded for your page.

---
While working on this I noticed another slightly related bug.  The "front page" template of the the 2024 theme contains a "query loop" and doesn't use "post title" and "post content" templates directly. This has a number of consequences that make the editor a bit different than the frontend.

 - Basically the query loop is showing "no results" found for me instead of showing the "fixed homepage" within the query. (I'm not entirely certain why is this happening)
 - The second issue is that we have a bug in the "template preview" option under the "template" menu. If you disable  "template preview", what we're trying to do is to guess which post blocks are used in the template and only show these in the editor, but the algorithm excludes all query block. Which means we end up with an "empty template" cc @ramonjd @andrewserong I think this second bug is probably something you can reproduce with the regular "page" template if you use a query block. (maybe even the first bug).

### CODE REVIEW

File: packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js [hunk 1]
```diff
@@ -54,7 +55,6 @@ function useResolveEditedEntityAndContext( { postId, postType } ) {
 			const {
 				getEditedEntityRecord,
 				getEntityRecords,
-				getDefaultTemplateId,
```
Comment (youknowriad) [inline packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js:57]: If this PR is considered good, I'm going to include the removal of this selector/resolver/reducer as well since it's something recent (a new API not in core) that is not needed anymore.

**Comment by jameskoster** [comment]
I can't comment on the code but I can confirm this is working correctly now 🙏 

> The "front page" template of the the 2024 theme contains a "query loop" and doesn't use "post title" and "post content" templates directly.

TT4 doesn't include a Front Page template by default. Perhaps the issue is caused by one of the patterns that is suggested when you create Front Page manually:

Screenshot 2023-11-20 at 15 39 04

**Comment by youknowriad** [comment]
> TT4 doesn't include a Front Page template by default. Perhaps the issue is caused by one of the patterns that is suggested when you create Front Page manually:

Yeah, I guess I had a lingering front page template, so maybe it's an issue with any "page" template that contain a "query"

File: packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js [hunk 2]
```diff
@@ -85,10 +89,10 @@ function useResolveEditedEntityAndContext( { postId, postType } ) {
...
 				}
 
-				// If no template is assigned, use the default template.
-				return getDefaultTemplateId( {
-					slug: `${ postTypeToResolve }-${ editedEntity?.slug }`,
-				} );
+				const path = getPathAndQueryString( editedEntity.link );
+
+				// If no template is assigned, use the link to fetch the right template.
+				return __experimentalGetTemplateForLink( path )?.id;
```
Comment (ramonjd) [inline packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js:95]: In my testing, this works well, i.e., the front page template is loaded in preference to the static home page in the reader settings.

However, I noticed that when opening the single page (Design > Pages > Static home page), the front page template is also loaded.

When on a page details view, should the page's template rather be loaded? The page's details in the left sidebar describe the page's meta but display's the front page template, which appears unrelated.

### Trunk

On the static page's home details view, we get the pages template, e.g., `twentytwentyfour//page`. 

Screenshot 2023-11-21 at 12 22 42 pm

### This PR
Here we're loading `twentytwentyfour//front-page`

Screenshot 2023-11-21 at 12 22 50 pm

What's the preferred way to handle page details? Should we still load the page's true template so folks can edit it as per other pages?

**Review by andrewserong — COMMENTED** [review]
Thanks for fixing this, this is testing better than on trunk for me 👍

| Before (uses page template) | After (uses front page template) |
| --- | --- |
| image | image |

One issue I ran into is that if I manually set the page itself to have a particular template set, then loading the site editor will still use that one, whereas the site frontend winds up using the Front page template. I think the cause is the early return if `currentTemplateSlug` is set?

| Site editor where custom template is shown as in use | Site frontend which uses front page template |
| --- | --- |
| image | image |

---

> The second issue is that we have a bug in the "template preview" option under the "template" menu. If you disable "template preview", what we're trying to do is to guess which post blocks are used in the template and only show these in the editor, but the algorithm excludes all query block.

I think the bug appears to be slightly more than that — if the template doesn't contain a Content block, then I still get dropped into the editor in the content editing mode, but without any blocks being editable. In order to edit the template we need to click on a block in the template and switch to the template mode. I.e. this is how it looks when I open the editor without a Content block — nothing in the list view, and nothing under the Content panel on the right:

image

So in the case that there are no content blocks outside of the query loop, perhaps one option could be to default to the template editing mode, and hide the Template Preview toggle entirely?

File: packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js (hunk 2 above)
Comment (youknowriad) [reply packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js:95]: I think the "front page" should load even in that case because that's the actual template that is used for that specific page anyway in the frontend.

**Comment by youknowriad** [comment]
@andrewserong 

> One issue I ran into is that if I manually set the page itself to have a particular template set, then loading the site editor will still use that one, whereas the site frontend winds up using the Front page template. I think the cause is the early return if currentTemplateSlug is set?

Interesting, it's weird that the frontend continue using the front page template in this case 🤔 even if the user explicitly set a custom template. There are two ways to go about this:

 - Either consider that a bug and fix the frontend.
 - Drop this PR and continue recreating the server side template resolution algorithm on the client.

The latter is probably safer as the former can also be seen as a breaking change but I also wonder if people raised this before on Core.

> So in the case that there are no content blocks outside of the query loop, perhaps one option could be to default to the template editing mode, and hide the Template Preview toggle entirely?

What about a fallback to "page title + page content" in this case.

File: packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js (hunk 2 above)
Comment (jameskoster) [reply packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js:95]: That's correct, Front Page will be used for the site homepage regardless.

**Comment by jameskoster** [comment]
> Interesting, it's weird that the frontend continue using the front page template in this case 🤔 even if the user explicitly set a custom template.

Afaik that's how the Front Page template has always worked. It will display the site homepage regardless of reading settings, or which template was originally applied to a static homepage. 

Related to this, it doesn't really make sense to allow the user to change the template applied to their static homepage when Front Page exists.

File: packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js (hunk 2 above)
Comment (ramonjd) [reply packages/edit-site/src/components/sync-state-with-url/use-init-edited-entity-from-url.js:95]: Thanks for the confirmation. 👍🏻 

I suppose I just found it confusing to see the page details on the left, including excerpt etc, and not the corresponding representation of that page on the right.

**Comment by andrewserong** [comment]
> continue recreating the server side template resolution algorithm on the client.

I think I lean a little in this direction, as changing how the behaviour works when folks set a front page under Reading could feel like a breaking change, as you mention.

> What about a fallback to "page title + page content" in this case.

That sounds like a better fallback if no blocks are present, good idea 👍

**Comment by youknowriad** [comment]
Thanks for the reviews folks, I've opened an alternative PR in #56418
